### PR TITLE
[SPARK-12854][SQL] Implement complex types support in ColumnarBatch

### PR DESCRIPTION
**Pull request by nongli**
This patch adds support for complex types for ColumnarBatch. ColumnarBatch supports structs
and arrays. There is a simple mapping between the richer catalyst types to these two. Strings
are treated as an array of bytes.

ColumnarBatch will contain a column for each node of the schema. Non-complex schemas consists
of just leaf nodes. Structs represent an internal node with one child for each field. Arrays
are internal nodes with one child. Structs just contain nullability. Arrays contain offsets
and lengths into the child array. This structure is able to handle arbitrary nesting. It has
the key property that we maintain columnar throughout and that primitive types are only stored
in the leaf nodes and contiguous across rows. For example, if the schema is 

```
array<array<int>>
```

There are three columns in the schema. The internal nodes each have one children. The leaf node contains all the int data stored consecutively.

As part of this, this patch adds append APIs in addition to the Put APIs (e.g. putLong(rowid, v)
vs appendLong(v)). These APIs are necessary when the batch contains variable length elements.
The vectors are not fixed length and will grow as necessary. This should make the usage a lot
simpler for the writer.

### CODE REVIEW

**Comment by rxin** [comment]
if the schema is array>,? array of int?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java [hunk 1]
```diff
@@ -17,22 +17,37 @@
...
 
 /**
  * This class represents a column of values and provides the main APIs to access the data
  * values. It supports all the types and contains get/put APIs as well as their batched versions.
  * The batched versions are preferable whenever possible.
  *
- * Most of the APIs take the rowId as a parameter. This is the local 0-based row id for values
+ * To handle nested schemas, ColumnVector has two types: Arrays and Structs. In both cases these
+ * columns have child columns. All of the data is stored in the child columns and the parent column
+ * contains nullability, and in the case of Arrays, the lengths and offsets into the child column.
```
Comment (rxin) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java]: can you explain how lengths and offsets are stored? also is there a single "parent" column that encodes nullability, length, and offset?

Comment (nongli) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java]: I'll update the comment when I rev the PR but the answer is:

I'm not sure what single "parent" means. The array is a column that stores nullability, lengths and offsets. The child column stores the values, including their nullability. Either one can be independently nullable or not.

Lengths and offsets are encoded like 'ints'.

**Comment by kiszk** [comment]
Will ColumnarBatch become a column version of UnsafeRow? Or, it will become more general columnar store?

**Comment by nongli** [comment]
@kiszk We're looking to make the execution engine represent data in a columnar way. I'm not sure what you mean by more general columnar store. I usually think of that more for storage.

File: unsafe/src/main/java/org/apache/spark/unsafe/Platform.java [hunk 2]
```diff
@@ -105,6 +105,17 @@ public static void freeMemory(long address) {
     _UNSAFE.freeMemory(address);
   }
 
+  public static long reallocateMemory(long address, long oldSize, long newSize) {
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/Platform.java]: Should this also work for on-heap?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OnHeapColumnVector.java [hunk 3]
```diff
@@ -173,4 +232,86 @@ public final void putDoubles(int rowId, int count, byte[] src, int srcIndex) {
...
+    if (requiredCapacity > capacity) reserveInternal(requiredCapacity * 2);
+  }
+
+  // Spilt this function out since it is the slow path.
+  private final void reserveInternal(int newCapacity) {
+    if (this.resultArray != null) {
+      int[] newLengths = new int[newCapacity];
+      int[] newOffsets = new int[newCapacity];
+      if (this.arrayLengths != null) {
+        System.arraycopy(this.arrayLengths, 0, newLengths, 0, elementsAppended);
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OnHeapColumnVector.java]: How does it compare to Unsafe.copyMemory()?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java [hunk 4]
```diff
@@ -82,25 +91,53 @@ public void close() {
...
+      return parent.numCols();
     }
 
     @Override
+    /**
+     * Revisit this. This is expensive.
+     */
     public final InternalRow copy() {
-      throw new NotImplementedException();
+      UnsafeRow row = new UnsafeRow(parent.numCols());
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: Should we re-use the row?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java [hunk 5]
```diff
@@ -82,25 +91,53 @@ public void close() {
...
 
     @Override
     public final int numFields() {
-      return ColumnarBatch.this.numCols();
+      return parent.numCols();
     }
 
     @Override
+    /**
+     * Revisit this. This is expensive.
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: This may be too slow for Join (or any operator that requires to hold a row). Could we use a lazy generated UnsafeProjection here?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java [hunk 6]
```diff
@@ -99,6 +185,26 @@ public void reset() {
   /**
    * Sets the value at rowId to `value`.
    */
+  public abstract void putByte(int rowId, byte value);
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java]: Should it also have putShort()/putFloat()?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java [hunk 7]
```diff
@@ -170,12 +204,13 @@ public final CalendarInterval getInterval(int ordinal) {
 
     @Override
     public final InternalRow getStruct(int ordinal, int numFields) {
-      throw new NotImplementedException();
+      return ColumnVectorUtils.toRow(parent.column(ordinal).getStruct(rowId));
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: Will this be too slow?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java [hunk 8]
```diff
@@ -170,12 +204,13 @@ public final CalendarInterval getInterval(int ordinal) {
...
     }
 
     @Override
     public final ArrayData getArray(int ordinal) {
-      throw new NotImplementedException();
+      ColumnVector.Array array = parent.column(ordinal).getArray(rowId);
+      return ColumnVectorUtils.toGenericArray(array);
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: Will this be too slow?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java (hunk 6 above)
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java]: We could separate these primitive types as another PR

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java [hunk 9]
```diff
@@ -42,13 +58,77 @@ public static ColumnVector allocate(int capacity, DataType type, MemoryMode mode
...
+  /**
+   * Holder object to return an array. This object is intended to be reused. Callers should
+   * copy the data out if it needs to be stored.
+   * TODO: consider adding all the get APIs that are 0 indexed. This would just be a convenience
+   * wrapper. Measure this for common usage patterns and see if there is an overhead.
+   *   int getInt(index) {
+   *     return data.getInt(index + offset);
+   *   }
+   */
+  public static final class Array {
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java]: Can we make this extend ArrayData? similar to what we have done for Row.

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java (hunk 6 above)
Comment (nongli) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java]: Yea, we're still missing those two types and probably decimal. I'll do those in a follow up.

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java (hunk 4 above)
Comment (nongli) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: Is that valid for copy? Copy seems like it can't have those semanatics. I think we need to remove copy() from the row interface internally. Or at least change it to take to something like:

void copy(UnsafeRowWriter writer). We should be able to directly write this to the destination.

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java (hunk 5 above)
Comment (nongli) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: Yea I think matei was looking at some code that would serve this purpose. I expect this too be too slow.

File: unsafe/src/main/java/org/apache/spark/unsafe/Platform.java (hunk 2 above)
Comment (nongli) [inline unsafe/src/main/java/org/apache/spark/unsafe/Platform.java]: I have this in the same group of functions as freeMemory, allocateMemory. What would the on heap version look like?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java (hunk 8 above)
Comment (nongli) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: I think for this and getStruct, it will be too slow. We are materializing the collection which we should avoid if possible. 

I think we need to clean up InternalRow's interface and then have getStruct return that.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchBenchmark.scala [hunk 10]
```diff
@@ -314,8 +338,60 @@ object ColumnarBatchBenchmark {
...
+        }
+        column.reset()
+      }
+    }
+
+    /*
+    String Read/Write:                       Avg Time(ms)    Avg Rate(M/s)  Relative Rate
+    -------------------------------------------------------------------------------------
+    On Heap                                         457.0            35.85         1.00 X
+    Off Heap                                       1206.0            13.59         0.38 X
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchBenchmark.scala]: utf8string supports reading offheap too so you can probably improve this substantially.

**Comment by rxin** [comment]
Going to merge this in master. Thanks.

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java [hunk 11]
```diff
@@ -166,12 +727,63 @@ public void reset() {
...
     this.type = type;
+
+    if (type instanceof ArrayType || type instanceof BinaryType || type instanceof StringType) {
+      DataType childType;
+      int childCapacity = capacity;
+      if (type instanceof ArrayType) {
+        childType = ((ArrayType)type).elementType();
+      } else {
+        childType = DataTypes.ByteType;
+        childCapacity *= DEFAULT_ARRAY_LENGTH;
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java]: Why only grow the capacity for non-array type?
